File: linkable-text/src/main/java/com/github/fobid/linkabletext/widget/LinkableTextView.java
package com.github.fobid.linkabletext.widget;

import android.annotation.TargetApi;
import android.content.Context;
import android.content.res.TypedArray;
import android.os.Build;
import android.support.annotation.Nullable;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.TextUtils;
import android.text.method.MovementMethod;
import android.text.style.URLSpan;
import android.text.util.Linkify;
import android.util.AttributeSet;
import android.util.Patterns;
import android.widget.TextView;

import com.github.fobid.linkabletext.R;
import com.github.fobid.linkabletext.annotation.LinkType;
import com.github.fobid.linkabletext.text.method.LinkableMovementMethod;
import com.github.fobid.linkabletext.view.LinkableCallback;
import com.github.fobid.linkabletext.view.OnLinkClickListener;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Created by partner on 2016-11-17.
 */

public class LinkableTextView extends TextView {

    public interface Link {
        int HASH_TAG = 0x1;
        int MENTION = 0x2;
        int EMAIL_ADDRESS = 0x3;
        int PHONE = 0x4;
        int WEB_URL = 0x5;
        int IP_ADDRESS = 0x6;
    }

    private static final Pattern IP_ADDRESS_PATTERN = Pattern.compile("^([01]?\\d\\d?|2[0-4]\\d|25[0-5])\\.([01]?\\d\\d?|2[0-4]\\d|25[0-5])\\.([01]?\\d\\d?|2[0-4]\\d|25[0-5])\\.([01]?\\d\\d?|2[0-4]\\d|25[0-5])$");
    private static Pattern MENTION_PATTERN;
    private static Pattern HASHTAG_PATTERN;

    private boolean enabledLinks = true;
    private boolean enabledHashtag = true;
    private boolean enabledMention = true;
    private boolean enabledEmailAddress = true;
    private boolean enabledPhone = true;
    private boolean enabledWebUrl = true;
    private boolean enabledIpAddress = true;

    private boolean enabledUnderlines = true;
    private boolean enabledHashtagUnderline = true;
    private boolean enabledMentionUnderline = true;
    private boolean enabledEmailAddressUnderline = true;
    private boolean enabledPhoneUnderline = true;
    private boolean enabledWebUrlUnderline = true;
    private boolean enabledIpAddressUnderline = true;

    public LinkableTextView(Context context) {
        super(context);
    }

    public LinkableTextView(Context context, AttributeSet attrs) {
        super(context, attrs);

        initView(context, attrs);
    }

    public LinkableTextView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        initView(context, attrs);
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public LinkableTextView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);

        initView(context, attrs);
    }

    private void initView(Context context, AttributeSet attrs) {
        String mentionPattern = null;
        String hashtagPattern = null;

        TypedArray a = context.obtainStyledAttributes(attrs, R.styleable.LinkableTextView);

        int n = a.getIndexCount();

        for (int i = 0; i < n; i++) {
            int attr = a.getIndex(i);

            if (attr == R.styleable.LinkableTextView_pattern_mention)
                mentionPattern = a.getString(attr);

            else if (attr == R.styleable.LinkableTextView_pattern_hashtag)
                hashtagPattern = a.getString(attr);

            else if (attr == R.styleable.LinkableTextView_enabledLinks)
                enabledLinks = a.getBoolean(attr, true);

            else if (attr == R.styleable.LinkableTextView_enabledHashtag)
                enabledHashtag = a.getBoolean(attr, true);

            else if (attr == R.styleable.LinkableTextView_enabledMention)
                enabledMention = a.getBoolean(attr, true);

            else if (attr == R.styleable.LinkableTextView_enabledEmail_address)
                enabledEmailAddress = a.getBoolean(attr, true);

            else if (attr == R.styleable.LinkableTextView_enabledPhone)
                enabledPhone = a.getBoolean(attr, true);

            else if (attr == R.styleable.LinkableTextView_enabledWeb_url)
                enabledWebUrl = a.getBoolean(attr, true);

            else if (attr == R.styleable.LinkableTextView_enabledIp_address)
                enabledIpAddress = a.getBoolean(attr, true);

            else if (attr == R.styleable.LinkableTextView_enabledUnderlines)
                enabledUnderlines = a.getBoolean(attr, true);

            else if (attr == R.styleable.LinkableTextView_enabledHashtagUnderline)
                enabledHashtagUnderline = a.getBoolean(attr, true);

            else if (attr == R.styleable.LinkableTextView_enabledMentionUnderline)
                enabledMentionUnderline = a.getBoolean(attr, true);

            else if (attr == R.styleable.LinkableTextView_enabledEmail_addressUnderline)
                enabledEmailAddressUnderline = a.getBoolean(attr, true);

            else if (attr == R.styleable.LinkableTextView_enabledPhoneUnderline)
                enabledPhoneUnderline = a.getBoolean(attr, true);

            else if (attr == R.styleable.LinkableTextView_enabledWeb_urlUnderline)
                enabledWebUrlUnderline = a.getBoolean(attr, true);

            else if (attr == R.styleable.LinkableTextView_enabledIp_addressUnderline)
                enabledIpAddressUnderline = a.getBoolean(attr, true);
        }
        setMentionPattern(mentionPattern);
        setHashtagPattern(hashtagPattern);

        setEnabledLinks(enabledLinks);
        setEnabledHashtag(enabledHashtag);
        setEnabledMention(enabledMention);
        setEnabledEmailAddress(enabledEmailAddress);
        setEnabledPhone(enabledPhone);
        setEnabledWebUrl(enabledWebUrl);
        setEnabledIpAddress(enabledIpAddress);

        setEnabledUnderlines(enabledUnderlines);
        setEnabledHashtagUnderline(enabledHashtagUnderline);
        setEnabledMentionUnderline(enabledMentionUnderline);
        setEnabledEmailAddressUnderline(enabledEmailAddressUnderline);
        setEnabledPhoneUnderline(enabledPhoneUnderline);
        setEnabledWebUrlUnderline(enabledWebUrlUnderline);
        setEnabledIpAddressUnderline(enabledIpAddressUnderline);

        a.recycle();
    }

    public void setMentionPattern(@Nullable String mentionPattern) {
        if (!TextUtils.isEmpty(mentionPattern)) {
            MENTION_PATTERN = Pattern.compile(mentionPattern);
        } else {
            MENTION_PATTERN = Pattern.compile("@([A-Za-z0-9_-]+)");
        }
    }

    public Pattern getMentionPattern() {
        return MENTION_PATTERN;
    }

    public void setDefaultMentionPattern() {
        setMentionPattern(null);
    }

    public void setHashtagPattern(@Nullable String hashTagPattern) {
        if (!TextUtils.isEmpty(hashTagPattern)) {
            HASHTAG_PATTERN = Pattern.compile(hashTagPattern);
        } else {
            HASHTAG_PATTERN = Pattern.compile("#([A-Za-z0-9_-]+)");
        }
    }

    public Pattern getHashtagPattern() {
        return HASHTAG_PATTERN;
    }

    public void setDefaultHashtagPattern() {
        setHashtagPattern(null);
    }

    public boolean isEnabledLinks() {
        return enabledLinks;
    }

    public void setEnabledLinks(boolean enabledLinks) {
        this.enabledLinks = enabledLinks;
    }

    public boolean isEnabledHashtag() {
        return enabledHashtag;
    }

    public void setEnabledHashtag(boolean enabledHashtag) {
        this.enabledHashtag = enabledHashtag;
    }

    public boolean isEnabledMention() {
        return enabledMention;
    }

    public void setEnabledMention(boolean enabledMention) {
        this.enabledMention = enabledMention;
    }

    public boolean isEnabledEmailAddress() {
        return enabledEmailAddress;
    }

    public void setEnabledEmailAddress(boolean enabledEmailAddress) {
        this.enabledEmailAddress = enabledEmailAddress;
    }

    public boolean isEnabledPhone() {
        return enabledPhone;
    }

    public void setEnabledPhone(boolean enabledPhone) {
        this.enabledPhone = enabledPhone;
    }

    public boolean isEnabledWebUrl() {
        return enabledWebUrl;
    }

    public void setEnabledWebUrl(boolean enabledWebUrl) {
        this.enabledWebUrl = enabledWebUrl;
    }

    public boolean isEnabledIpAddress() {
        return enabledIpAddress;
    }

    public void setEnabledIpAddress(boolean enabledIpAddress) {
        this.enabledIpAddress = enabledIpAddress;
    }

    public boolean isEnabledUnderlines() {
        return enabledUnderlines;
    }

    public void setEnabledUnderlines(boolean enabledUnderlines) {
        this.enabledUnderlines = enabledUnderlines;
    }

    public boolean isEnabledHashtagUnderline() {
        return enabledHashtagUnderline;
    }

    public void setEnabledHashtagUnderline(boolean enabledHashtagUnderline) {
        this.enabledHashtagUnderline = enabledHashtagUnderline;
    }

    public boolean isEnabledMentionUnderline() {
        return enabledMentionUnderline;
    }

    public void setEnabledMentionUnderline(boolean enabledMentionUnderline) {
        this.enabledMentionUnderline = enabledMentionUnderline;
    }

    public boolean isEnabledEmailAddressUnderline() {
        return enabledEmailAddressUnderline;
    }

    public void setEnabledEmailAddressUnderline(boolean enabledEmailAddressUnderline) {
        this.enabledEmailAddressUnderline = enabledEmailAddressUnderline;
    }

    public boolean isEnabledPhoneUnderline() {
        return enabledPhoneUnderline;
    }

    public void setEnabledPhoneUnderline(boolean enabledPhoneUnderline) {
        this.enabledPhoneUnderline = enabledPhoneUnderline;
    }

    public boolean isEnabledWebUrlUnderline() {
        return enabledWebUrlUnderline;
    }

    public void setEnabledWebUrlUnderline(boolean enabledWebUrlUnderline) {
        this.enabledWebUrlUnderline = enabledWebUrlUnderline;
    }

    public boolean isEnabledIpAddressUnderline() {
        return enabledIpAddressUnderline;
    }

    public void setEnabledIpAddressUnderline(boolean enabledIpAddressUnderline) {
        this.enabledIpAddressUnderline = enabledIpAddressUnderline;
    }

    public void setOnLinkClickListener(@Nullable final OnLinkClickListener listener) {
        setOnLinkClickListener(new LinkableCallback() {
            @Override
            public void onMatch(@LinkType int type, String value) {
                if (listener == null || !enabledLinks)
                    return;

                switch (type) {
                    case Link.HASH_TAG:
                        if (enabledHashtag)
                            listener.onHashtagClick(value);
                        break;

                    case Link.MENTION:
                        if (enabledMention)
                            listener.onMentionClick(value);
                        break;

                    case Link.EMAIL_ADDRESS:
                        if (enabledEmailAddress)
                            listener.onEmailAddressClick(value);
                        break;
                    case Link.IP_ADDRESS:
                        if (!enabledIpAddress)
                            break;
                    case Link.WEB_URL:
                        if (enabledWebUrl)
                            listener.onWebUrlClick(value);
                        break;

                    case Link.PHONE:
                        if (enabledPhone)
                            listener.onPhoneClick(value);
                        break;
                }
            }
        });
    }

    public void setOnLinkClickListener(@Nullable final LinkableCallback callback) {
        Linkify.TransformFilter filter = new Linkify.TransformFilter() {
            public final String transformUrl(final Matcher match, String url) {
                return match.group();
            }
        };

        if (enabledLinks) {
            if (enabledUnderlines) {
                if (enabledHashtag) {
                    Linkify.addLinks(this, HASHTAG_PATTERN,
                            LinkableMovementMethod.LINKABLE_HASHTAG_SCHEME, null, filter);

                    if (!enabledHashtagUnderline)
                        stripUnderlines();
                }

                if (enabledMention) {
                    Linkify.addLinks(this, MENTION_PATTERN,
                            LinkableMovementMethod.LINKABLE_MENTION_SCHEME, null, filter);

                    if (!enabledMentionUnderline)
                        stripUnderlines();
                }

                if (enabledEmailAddress) {
                    Linkify.addLinks(this, Patterns.EMAIL_ADDRESS, null, null, filter);

                    if (!enabledEmailAddressUnderline)
                        stripUnderlines();
                }

                if (enabledPhone) {
                    Linkify.addLinks(this, Patterns.PHONE, null, null, filter);

                    if (!enabledPhoneUnderline)
                        stripUnderlines();
                }

                if (enabledWebUrl) {
                    Linkify.addLinks(this, Patterns.WEB_URL, null, null, filter);

                    if (!enabledWebUrlUnderline)
                        stripUnderlines();
                }

                if (enabledIpAddress) {
                    Linkify.addLinks(this, IP_ADDRESS_PATTERN,
                            LinkableMovementMethod.LINKABLE_IP_ADDRESS_SCHEME, null, filter);

                    if (!enabledIpAddressUnderline)
                        stripUnderlines();
                }
            } else {
                if (enabledHashtag)
                    Linkify.addLinks(this, HASHTAG_PATTERN,
                            LinkableMovementMethod.LINKABLE_HASHTAG_SCHEME, null, filter);

                if (enabledMention)
                    Linkify.addLinks(this, MENTION_PATTERN,
                            LinkableMovementMethod.LINKABLE_MENTION_SCHEME, null, filter);

                if (enabledEmailAddress)
                    Linkify.addLinks(this, Patterns.EMAIL_ADDRESS, null, null, filter);

                if (enabledPhone)
                    Linkify.addLinks(this, Patterns.PHONE, null, null, filter);

                if (enabledWebUrl)
                    Linkify.addLinks(this, Patterns.WEB_URL, null, null, filter);

                if (enabledIpAddress)
                    Linkify.addLinks(this, IP_ADDRESS_PATTERN,
                            LinkableMovementMethod.LINKABLE_IP_ADDRESS_SCHEME, null, filter);

                stripUnderlines();
            }
        }

        MovementMethod movementMethod = null;
        if (callback != null) {
            movementMethod = new LinkableMovementMethod(callback);
        }
        setMovementMethod(movementMethod);
    }

    private void stripUnderlines() {
        Spannable s = new SpannableString(getText());
        URLSpan[] spans = s.getSpans(0, s.length(), URLSpan.class);
        for (URLSpan span : spans) {
            int start = s.getSpanStart(span);
            int end = s.getSpanEnd(span);
            s.removeSpan(span);
            span = new URLSpanNoUnderline(span.getURL());
            s.setSpan(span, start, end, 0);
        }
        setText(s);
    }

}

File: linkable-text/src/main/java/com/github/fobid/linkabletext/widget/LinkableTextView.kt
package com.github.fobid.linkabletext.widget

import android.annotation.TargetApi
import android.content.Context
import android.os.Build
import android.text.SpannableString
import android.text.method.MovementMethod
import android.text.style.URLSpan
import android.text.util.Linkify
import android.util.AttributeSet
import android.util.Patterns
import androidx.appcompat.widget.AppCompatTextView
import com.github.fobid.linkabletext.R
import com.github.fobid.linkabletext.text.method.LinkableMovementMethod
import com.github.fobid.linkabletext.view.LinkableCallback
import com.github.fobid.linkabletext.view.OnLinkClickListener
import java.util.regex.Pattern

@TargetApi(Build.VERSION_CODES.LOLLIPOP)
class LinkableTextView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : AppCompatTextView(context, attrs, defStyleAttr) {

    object Link {
        const val HASH_TAG = 0x1
        const val MENTION = 0x2
        const val EMAIL_ADDRESS = 0x3
        const val PHONE = 0x4
        const val WEB_URL = 0x5
        const val IP_ADDRESS = 0x6
    }

    private var enabledLinks = true
    private var enabledHashtag = true
    private var enabledMention = true
    private var enabledEmailAddress = true
    private var enabledPhone = true
    private var enabledWebUrl = true
    private var enabledIpAddress = true

    private var enabledUnderlines = true
    private var enabledHashtagUnderline = true
    private var enabledMentionUnderline = true
    private var enabledEmailAddressUnderline = true
    private var enabledPhoneUnderline = true
    private var enabledWebUrlUnderline = true
    private var enabledIpAddressUnderline = true

    init {
        var mentionPattern = ""
        var hashtagPattern = ""

        context.obtainStyledAttributes(attrs, R.styleable.LinkableTextView)
            .also {
                for (i in 0 until it.indexCount) {

                    when (val attr = it.getIndex(i)) {
                        R.styleable.LinkableTextView_pattern_mention ->
                            mentionPattern = it.getString(attr) ?: ""
                        R.styleable.LinkableTextView_pattern_hashtag ->
                            hashtagPattern = it.getString(attr) ?: ""
                        R.styleable.LinkableTextView_enabledLinks ->
                            enabledLinks = it.getBoolean(attr, true)
                        R.styleable.LinkableTextView_enabledHashtag ->
                            enabledHashtag = it.getBoolean(attr, true)
                        R.styleable.LinkableTextView_enabledMention ->
                            enabledMention = it.getBoolean(attr, true)
                        R.styleable.LinkableTextView_enabledEmail_address ->
                            enabledEmailAddress = it.getBoolean(attr, true)
                        R.styleable.LinkableTextView_enabledPhone ->
                            enabledPhone = it.getBoolean(attr, true)
                        R.styleable.LinkableTextView_enabledPhone ->
                            enabledPhone = it.getBoolean(attr, true)
                        R.styleable.LinkableTextView_enabledWeb_url ->
                            enabledWebUrl = it.getBoolean(attr, true)
                        R.styleable.LinkableTextView_enabledIp_address ->
                            enabledIpAddress = it.getBoolean(attr, true)
                        R.styleable.LinkableTextView_enabledUnderlines ->
                            enabledUnderlines = it.getBoolean(attr, true)
                        R.styleable.LinkableTextView_enabledHashtagUnderline ->
                            enabledHashtagUnderline = it.getBoolean(attr, true)
                        R.styleable.LinkableTextView_enabledMentionUnderline ->
                            enabledMentionUnderline = it.getBoolean(attr, true)
                        R.styleable.LinkableTextView_enabledEmail_addressUnderline ->
                            enabledEmailAddressUnderline = it.getBoolean(attr, true)
                        R.styleable.LinkableTextView_enabledPhoneUnderline ->
                            enabledPhoneUnderline = it.getBoolean(attr, true)
                        R.styleable.LinkableTextView_enabledWeb_urlUnderline ->
                            enabledWebUrlUnderline = it.getBoolean(attr, true)
                        R.styleable.LinkableTextView_enabledIp_addressUnderline ->
                            enabledIpAddressUnderline = it.getBoolean(attr, true)
                    }
                }

                setMentionPattern(mentionPattern)
                setHashtagPattern(hashtagPattern)

                setEnabledLinks(enabledLinks)
                setEnabledHashtag(enabledHashtag)
                setEnabledMention(enabledMention)
                setEnabledEmailAddress(enabledEmailAddress)
                setEnabledPhone(enabledPhone)
                setEnabledWebUrl(enabledWebUrl)
                setEnabledIpAddress(enabledIpAddress)

                setEnabledUnderlines(enabledUnderlines)
                setEnabledHashtagUnderline(enabledHashtagUnderline)
                setEnabledMentionUnderline(enabledMentionUnderline)
                setEnabledEmailAddressUnderline(enabledEmailAddressUnderline)
                setEnabledPhoneUnderline(enabledPhoneUnderline)
                setEnabledWebUrlUnderline(enabledWebUrlUnderline)
                setEnabledIpAddressUnderline(enabledIpAddressUnderline)
            }.apply {
                recycle()
            }
    }

    fun setMentionPattern(mentionPattern: String) {
        MENTION_PATTERN = Pattern.compile(
            when (mentionPattern.isNotEmpty()) {
                true -> mentionPattern
                else -> "@([A-Za-z0-9_-]+)"
            }
        )
    }

    fun setHashtagPattern(hashtagPattern: String) {
        HASHTAG_PATTERN = Pattern.compile(
            when (hashtagPattern.isNotEmpty()) {
                true -> hashtagPattern
                else -> "#(\\w+)"
            }
        )
    }

    fun setEnabledLinks(enabledLinks: Boolean) {
        this.enabledLinks = enabledLinks
    }

    fun setEnabledHashtag(enabledHashtag: Boolean) {
        this.enabledHashtag = enabledHashtag
    }

    fun setEnabledMention(enabledMention: Boolean) {
        this.enabledMention = enabledMention
    }

    fun setEnabledEmailAddress(enabledEmailAddress: Boolean) {
        this.enabledEmailAddress = enabledEmailAddress
    }

    fun setEnabledPhone(enabledPhone: Boolean) {
        this.enabledPhone = enabledPhone
    }

    fun setEnabledWebUrl(enabledWebUrl: Boolean) {
        this.enabledWebUrl = enabledWebUrl
    }

    fun setEnabledIpAddress(enabledIpAddress: Boolean) {
        this.enabledIpAddress = enabledIpAddress
    }

    fun setEnabledUnderlines(enabledUnderlines: Boolean) {
        this.enabledUnderlines = enabledUnderlines
    }

    fun setEnabledHashtagUnderline(enabledHashtagUnderline: Boolean) {
        this.enabledHashtagUnderline = enabledHashtagUnderline
    }

    fun setEnabledMentionUnderline(enabledMentionUnderline: Boolean) {
        this.enabledMentionUnderline = enabledMentionUnderline
    }

    fun setEnabledEmailAddressUnderline(enabledEmailAddressUnderline: Boolean) {
        this.enabledEmailAddressUnderline = enabledEmailAddressUnderline
    }

    fun setEnabledPhoneUnderline(enabledPhoneUnderline: Boolean) {
        this.enabledPhoneUnderline = enabledPhoneUnderline
    }

    fun setEnabledWebUrlUnderline(enabledWebUrlUnderline: Boolean) {
        this.enabledWebUrlUnderline = enabledWebUrlUnderline
    }

    fun setEnabledIpAddressUnderline(enabledIpAddressUnderline: Boolean) {
        this.enabledIpAddressUnderline = enabledIpAddressUnderline
    }

    fun setOnLinkClickListener(listener: OnLinkClickListener?) =
        setOnLinkClickListener(object : LinkableCallback {
            override fun onMatch(type: Int, value: String) {
                listener?.let {
                    if (!enabledLinks) {
                        return
                    }

                    when (type) {
                        Link.HASH_TAG ->
                            if (enabledHashtag) {
                                listener.onHashtagClick(value)
                            }
                        Link.MENTION ->
                            if (enabledMention) {
                                listener.onMentionClick(value)
                            }
                        Link.EMAIL_ADDRESS ->
                            if (enabledEmailAddress) {
                                listener.onEmailAddressClick(value)
                            }
                        Link.IP_ADDRESS ->
                            if (enabledIpAddress) {
                                listener.onWebUrlClick(value)
                            }
                        Link.WEB_URL ->
                            if (enabledWebUrl) {
                                listener.onWebUrlClick(value)
                            }
                        Link.PHONE ->
                            if (enabledPhone) {
                                listener.onPhoneClick(value)
                            }
                    }
                }
            }
        })

    fun setOnLinkClickListener(callback: LinkableCallback?) {
        val filter = Linkify.TransformFilter { match, _ -> match.group() }

        if (enabledLinks) {
            if (enabledUnderlines) {
                if (enabledHashtag) {
                    Linkify.addLinks(
                        this,
                        HASHTAG_PATTERN,
                        LinkableMovementMethod.LINKABLE_HASHTAG_SCHEME,
                        null,
                        filter
                    )

                    if (!enabledHashtagUnderline)
                        stripUnderlines()
                }

                if (enabledMention) {
                    Linkify.addLinks(
                        this,
                        MENTION_PATTERN,
                        LinkableMovementMethod.LINKABLE_MENTION_SCHEME,
                        null,
                        filter
                    )

                    if (!enabledMentionUnderline)
                        stripUnderlines()
                }

                if (enabledEmailAddress) {
                    Linkify.addLinks(this, Patterns.EMAIL_ADDRESS, null, null, filter)

                    if (!enabledEmailAddressUnderline)
                        stripUnderlines()
                }

                if (enabledPhone) {
                    Linkify.addLinks(this, Patterns.PHONE, null, null, filter)

                    if (!enabledPhoneUnderline)
                        stripUnderlines()
                }

                if (enabledWebUrl) {
                    Linkify.addLinks(this, Patterns.WEB_URL, null, null, filter)

                    if (!enabledWebUrlUnderline)
                        stripUnderlines()
                }

                if (enabledIpAddress) {
                    Linkify.addLinks(
                        this,
                        IP_ADDRESS_PATTERN,
                        LinkableMovementMethod.LINKABLE_IP_ADDRESS_SCHEME,
                        null,
                        filter
                    )

                    if (!enabledIpAddressUnderline)
                        stripUnderlines()
                }
            } else {
                if (enabledHashtag)
                    Linkify.addLinks(
                        this,
                        HASHTAG_PATTERN,
                        LinkableMovementMethod.LINKABLE_HASHTAG_SCHEME,
                        null,
                        filter
                    )

                if (enabledMention)
                    Linkify.addLinks(
                        this,
                        MENTION_PATTERN,
                        LinkableMovementMethod.LINKABLE_MENTION_SCHEME,
                        null,
                        filter
                    )

                if (enabledEmailAddress)
                    Linkify.addLinks(this, Patterns.EMAIL_ADDRESS, null, null, filter)

                if (enabledPhone)
                    Linkify.addLinks(this, Patterns.PHONE, null, null, filter)

                if (enabledWebUrl)
                    Linkify.addLinks(this, Patterns.WEB_URL, null, null, filter)

                if (enabledIpAddress)
                    Linkify.addLinks(
                        this,
                        IP_ADDRESS_PATTERN,
                        LinkableMovementMethod.LINKABLE_IP_ADDRESS_SCHEME,
                        null,
                        filter
                    )

                stripUnderlines()
            }
        }

        var movementMethod: MovementMethod? = null
        if (callback != null) {
            movementMethod = LinkableMovementMethod(callback)
        }
        setMovementMethod(movementMethod)
    }

    private fun stripUnderlines() {
        val spannableString = SpannableString(text)

        spannableString.getSpans(0, spannableString.length, URLSpan::class.java).let { spans ->
            for (span in spans) {
                val start = spannableString.getSpanStart(span)
                val end = spannableString.getSpanEnd(span)
                spannableString.removeSpan(span)
                spannableString.setSpan(URLSpanNoUnderline(span.url), start, end, 0)
            }
        }

        text = spannableString
    }

    companion object {
        private val IP_ADDRESS_PATTERN: Pattern =
            Pattern.compile("^([01]?\\d\\d?|2[0-4]\\d|25[0-5])\\.([01]?\\d\\d?|2[0-4]\\d|25[0-5])\\.([01]?\\d\\d?|2[0-4]\\d|25[0-5])\\.([01]?\\d\\d?|2[0-4]\\d|25[0-5])$")
        private lateinit var MENTION_PATTERN: Pattern
        private lateinit var HASHTAG_PATTERN: Pattern
    }
}